File: exercism/python/bob/bob.py
import re


def respond_to_question(s):
    return 'Sure.' if s.endswith('?') else None


def respond_to_noword(s):
    return "Fine. Be that way!" if s == '' else None


def respond_to_yelling(s):
    if s.isupper():
        if s[-1] == '?':
            return 'Calm down, I know what I\'m doing!'
        else:
            return 'Whoa, chill out!'
    else:
        return None


def hey(phrase):
    stripped = phrase.strip()
    return (respond_to_noword(stripped)
            or respond_to_yelling(stripped)
            or respond_to_question(stripped)
            or 'Whatever.')

File: exercism/scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = {
    val alphabit = ('a' to 'z') ++ ('A' to 'Z') 
    def isQuestion(str: String) = str.endsWith("?")
    def isYelling(str: String) = str.exists(alphabit.toSet) && str.forall(c => c.isUpper || !c.isLetter)

    statement.trim match {
      case str if str.isEmpty => "Fine. Be that way!"
      case str if isQuestion(str) && isYelling(str) => "Calm down, I know what I'm doing!"
      case str if isQuestion(str) => "Sure."
      case str if isYelling(str) => "Whoa, chill out!"
      case _ => "Whatever."
    }
  }
}
